concordium-node/src/grpc2.rs

```diff
@@ -2290,3 +2296,136 @@ impl<A> Require<tonic::Status> for Option<A> {
+    type Future = futures::future::BoxFuture<'static, Result<Self::Response, Self::Error>>;
+    type Response = S::Response;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: hyper::Request<hyper::Body>) -> Self::Future {
+        // See https://docs.rs/tower/latest/tower/trait.Service.html#be-careful-when-cloning-inner-services
+        // for details on why this is necessary
+        let clone = self.inner.clone();
+        let mut inner = std::mem::replace(&mut self.inner, clone);
+
+        let grpc_request_duration = self.stats.grpc_request_response_time.clone();
+
+        Box::pin(async move {
+            let endpoint_name = req.uri().path().to_owned();
+            let request_received = tokio::time::Instant::now();
+
+            let (mut response, duration) = {
+                // Forward the request.
+                let result = inner.call(req).await;
+                // Time taken for the inner service to send back a response, meaning for
+                // streaming gRPC methods this is the duration for it to first return a stream.
+                let duration = request_received.elapsed().as_secs_f64();
+                if result.is_err() {
+                    grpc_request_duration
+                        .with_label_values(&[endpoint_name.as_str(), "internal"])
+                        .observe(duration);
+                }
+                (result?, duration)
+            };
+
+            // Check if the gRPC status header is part of the HTTP headers, if not check for
+            // the status in the HTTP trailers.
+            // Every gRPC response are expected to provide the grpc-status in either the
+            // header or the trailer.
+            let grpc_status_header = response.headers().get("grpc-status");
+            if let Some(header_value) = grpc_status_header {
+                let status_code_label =
+                    get_grpc_code_label(tonic::Code::from_bytes(header_value.as_bytes()));
+
+                grpc_request_duration
+                    .with_label_values(&[endpoint_name.as_str(), status_code_label])
+                    .observe(duration);
+            } else if let Ok(Some(trailers)) =
+                hyper::body::HttpBody::trailers(response.body_mut()).await
+            {
+                if let Some(header_value) = trailers.get("grpc-status") {
+                    let status_code_label =
+                        get_grpc_code_label(tonic::Code::from_bytes(header_value.as_bytes()));
+
+                    grpc_request_duration
+                        .with_label_values(&[endpoint_name.as_str(), status_code_label])
+                        .observe(duration);
+                }
+            } else {
+                grpc_request_duration
+                    .with_label_values(&[endpoint_name.as_str(), "unknown"])
```

Comment:
Maybe it'd be better to do `get_grpc_code_label(Code::Unknown)` instead of the string `unknown` in case we ever change this.